kernel/src/process_loading.rs

```diff
@@ -314,191 +205,749 @@ fn load_processes_from_flash<C: Chip>(
+    ///
+    /// - They both have the same AppID, or
+    /// - They both have the same ShortId
+    ///
+    /// Since `process` is already loaded, we only have to enforce the AppID and
+    /// ShortId uniqueness guarantees.
+    fn is_blocked_from_loading_by_process(
+        &self,
+        pb: &ProcessBinary,
+        process: &dyn Process,
+    ) -> bool {
+        let same_app_id = self.policy.map_or(false, |policy| {
+            !policy.different_identifier_process(pb, process)
+        });
+        let same_short_app_id = self.policy.map_or(false, |policy| {
+            policy.to_short_id(pb) == process.short_app_id()
+        });
+
+        let blocks = same_app_id || same_short_app_id;
+
+        if config::CONFIG.debug_process_credentials {
+            debug!(
+                "Loading: Process {}({:#02x}) does{} block {}({:#02x})",
+                process.get_process_name(),
+                process.get_addresses().flash_start,
+                if blocks { " not" } else { "" },
+                pb.header.get_package_name().unwrap_or(""),
+                pb.flash.as_ptr() as usize,
+            );
+        }
+
+        blocks
+    }
+}
+
+impl<'a, C: Chip> ProcessLoadingAsync<'a> for SequentialProcessLoaderMachine<'a, C> {
+    fn set_client(&self, client: &'a dyn ProcessLoadingAsyncClient) {
+        self.client.set(client);
+    }
+
+    fn set_policy(&self, policy: &'a dyn AppIdPolicy) {
+        self.policy.replace(policy);
+    }
+
+    fn start(&self) {
+        self.state
+            .set(SequentialProcessLoaderMachineState::DiscoverProcessBinaries);
+        // Start an asynchronous flow so we can issue a callback on error.
+        self.deferred_call.set();
+    }
+}
+
+impl<'a, C: Chip> DeferredCallClient for SequentialProcessLoaderMachine<'a, C> {
+    fn handle_deferred_call(&self) {
+        // We use deferred calls to start the operation in the async loop.
+        match self.state.get() {
+            Some(SequentialProcessLoaderMachineState::DiscoverProcessBinaries) => {
+                self.load_and_check();
+            }
+            Some(SequentialProcessLoaderMachineState::LoadProcesses) => {
+                let ret = self.load_process_objects();
+                match ret {
+                    Ok(()) => {}
+                    Err(()) => {
+                        // If this failed for some reason, we still need to
+                        // signal that process loading has finished.
+                        self.client.map(|client| {
+                            client.process_loading_finished();
+                        });
+                    }
+                }
+            }
+            None => {}
+        }
+    }
+
+    fn register(&'static self) {
+        self.deferred_call.register(self);
+    }
+}
+
+impl<'a, C: Chip> crate::process_checker::ProcessCheckerMachineClient
+    for SequentialProcessLoaderMachine<'a, C>
+{
+    fn done(
+        &self,
+        process_binary: ProcessBinary,
```

Comment:
I am a little worried about this being passed by value -- should I be?

Reply:
I don't think so. It's "moved" not copied